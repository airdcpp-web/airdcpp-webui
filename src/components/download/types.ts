import * as UI from 'types/ui';


export type BrowseHandler = (() => void) | undefined;

export type PathDownloadHandler = (targetPath: string, targetFilename?: string) => Promise<any>;

export type LayoutProps = React.PropsWithChildren<{
  menuItems: React.ReactNode[];
  title: string;
  handleBrowse: BrowseHandler;
  t: UI.TranslateF;
}>;
