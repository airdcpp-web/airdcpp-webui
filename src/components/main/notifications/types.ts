export type NotificationLevel = 'error' | 'warning' | 'info' | 'success';
