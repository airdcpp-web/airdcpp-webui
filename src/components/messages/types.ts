import { IntersectionObserverProps } from 'react-intersection-observer';

import * as UI from '@/types/ui';

export interface CommonMessageProps extends Omit<IntersectionObserverProps, 'children'> {
  highlightMenuProps: UI.MessageActionMenuData;
}
