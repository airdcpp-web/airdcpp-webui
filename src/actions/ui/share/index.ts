export * from './ShareActions';
export * from './ShareExcludeActions';
export * from './ShareProfileActions';
export * from './ShareRootActions';
