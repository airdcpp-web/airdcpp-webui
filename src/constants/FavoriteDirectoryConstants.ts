const MODULE_URL = 'favorite_directories';
export default {
  MODULE_URL: MODULE_URL,

  GROUPED_DIRECTORIES_URL: MODULE_URL + '/grouped_paths',
  DIRECTORIES_URL: MODULE_URL,

  DIRECTORIES_UPDATED: 'favorite_directories_updated',
};
