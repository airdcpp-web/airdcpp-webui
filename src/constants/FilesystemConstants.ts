const MODULE_URL = 'filesystem';
export default {
  MODULE_URL: MODULE_URL,
  LIST_URL: MODULE_URL + '/list_items',
  DIRECTORY_URL: MODULE_URL + '/directory',
  DISK_INFO_URL: MODULE_URL + '/disk_info',

  LOCATION_DOWNLOAD: 'download',
  LOCATION_GENERIC: 'settings',
};
