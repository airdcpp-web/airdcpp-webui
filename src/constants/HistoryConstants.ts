const MODULE_URL = 'histories';

export enum HistoryStringEnum {
  SEARCH = 'search_pattern',
  EXCLUDE = 'search_excluded',
  DOWNLOAD_DIR = 'download_target',
}

export enum HistoryEntryEnum {
  HUB = 'hub',
  PRIVATE_CHAT = 'private_chat',
  FILELIST = 'filelist',
}

export default {
  MODULE_URL: MODULE_URL,

  STRINGS_URL: MODULE_URL + '/strings',
  SESSIONS_URL: MODULE_URL + '/sessions',
};
