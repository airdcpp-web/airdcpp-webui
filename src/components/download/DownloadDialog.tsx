//import PropTypes from 'prop-types';
// import { useRef } from 'react';

import * as React from 'react';
import { Route, Routes, useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import ShareConstants from 'constants/ShareConstants';
import {
  default as HistoryConstants,
  HistoryStringEnum,
} from 'constants/HistoryConstants';
import FavoriteDirectoryConstants from 'constants/FavoriteDirectoryConstants';
import FilesystemConstants from 'constants/FilesystemConstants';
import IconConstants from 'constants/IconConstants';

import NotificationActions from 'actions/NotificationActions';
import LoginStore from 'stores/LoginStore';

import DataProviderDecorator, {
  DataProviderDecoratorChildProps,
} from 'decorators/DataProviderDecorator';
import ModalRouteDecorator, {
  ModalRouteDecoratorChildProps,
} from 'decorators/ModalRouteDecorator';

import Modal, { ModalProps } from 'components/semantic/Modal';
import { FileBrowserDialog } from 'components/filebrowser';

import { runBackgroundSocketAction } from 'utils/ActionUtils';
import { toI18nKey, translate } from 'utils/TranslationUtils';
import { addHistory } from 'services/api/HistoryApi';

import { DownloadLayout, DownloadDataProps } from './layout';

import * as API from 'types/api';
import * as UI from 'types/ui';
import { PathDownloadHandler } from './types';

import './style.css';

export type DownloadDialogProps<
  ItemT extends UI.DownloadableItemInfo = UI.DownloadableItemInfo,
> = UI.ItemDownloadHandler<ItemT, Props<ItemT>>;

/*type DownloadItemIdType = string;

interface RouteProps {
  downloadItemId: DownloadItemIdType;
}*/

type DownloadDialogRouteProps = ModalRouteDecoratorChildProps;

interface DownloadDialogDataProps<
  ItemT extends UI.DownloadableItemInfo = UI.DownloadableItemInfo,
> extends DataProviderDecoratorChildProps,
    DownloadDataProps<ItemT> {}

type Props<ItemT extends UI.DownloadableItemInfo = UI.DownloadableItemInfo> =
  DownloadDialogProps<ItemT> & DownloadDialogDataProps<ItemT> & DownloadDialogRouteProps;

const DownloadDialog: React.FC<Props> = (props) => {
  const { t } = useTranslation();
  const navigate = useNavigate();

  const {
    downloadHandler,
    itemInfo,
    userGetter,
    params,
    session,
    historyPaths,
    favoritePaths,
    sharePaths,
    handleClose,
    ...other
  } = props;

  const handleDownload: PathDownloadHandler = async (targetPath, targetFilename) => {
    try {
      await downloadHandler(
        itemInfo,
        !!userGetter ? userGetter(params.downloadItemId!, props) : undefined,
        {
          target_name: !!targetFilename ? targetFilename : itemInfo.name,
          target_directory: targetPath,
          priority: API.QueuePriorityEnum.DEFAULT,
        },
        session,
      );
    } catch (e) {
      NotificationActions.error({
        title: t(toI18nKey('queueingFailed', UI.Modules.COMMON), {
          defaultValue: 'Failed to queue the item {{item.name}}',
          replace: {
            item: itemInfo,
          },
        }),
        message: e.message,
      });
    }

    runBackgroundSocketAction(
      () => addHistory(HistoryStringEnum.DOWNLOAD_DIR, targetPath),
      t,
    );

    handleClose();
  };

  const getInitialBrowsePath = () => {
    const path = historyPaths.length > 0 ? historyPaths[historyPaths.length - 1] : '';
    return itemInfo.type.id === 'directory' ? path : path + itemInfo.name;
  };

  const handleBrowse = () => {
    navigate(`browse`);
  };

  const commonDialogProps: ModalProps = {
    subHeader: itemInfo.name,
    title: translate('Download', t, UI.Modules.COMMON),
    icon: IconConstants.DOWNLOAD,
    closable: false,
  };

  const hasFileBrowserAccess = LoginStore.hasAccess(API.AccessEnum.FILESYSTEM_VIEW);
  return (
    <Routes>
      <Route
        path={`browse`}
        element={
          <FileBrowserDialog
            onConfirm={(path, directoryPath, fileName) =>
              handleDownload(directoryPath, fileName)
            }
            initialPath={getInitialBrowsePath()}
            selectMode={
              itemInfo.type.id === 'directory'
                ? UI.FileSelectModeEnum.DIRECTORY
                : UI.FileSelectModeEnum.FILE
            }
            historyId={FilesystemConstants.LOCATION_DOWNLOAD}
            approveCaption={translate('Download', t, UI.Modules.COMMON)}
            {...commonDialogProps}
          />
        }
      />
      <Route
        index
        element={
          <Modal
            className="download-dialog"
            fullHeight={true}
            {...commonDialogProps}
            {...other}
          >
            <DownloadLayout
              downloadHandler={handleDownload}
              handleBrowse={hasFileBrowserAccess ? handleBrowse : undefined}
              historyPaths={historyPaths}
              favoritePaths={favoritePaths}
              sharePaths={sharePaths}
              itemInfo={itemInfo}
            />
          </Modal>
        }
      />
    </Routes>
  );
};

export default ModalRouteDecorator<DownloadDialogProps>(
  DataProviderDecorator<
    DownloadDialogProps & DownloadDialogRouteProps,
    DownloadDialogDataProps
  >(DownloadDialog, {
    urls: {
      sharePaths: ShareConstants.GROUPED_ROOTS_GET_URL,
      favoritePaths: FavoriteDirectoryConstants.GROUPED_DIRECTORIES_URL,
      historyPaths: HistoryConstants.STRINGS_URL + '/' + HistoryStringEnum.DOWNLOAD_DIR,
      itemInfo: ({ params, itemDataGetter }, socket) => {
        return itemDataGetter(params.downloadItemId!, socket);
      },
    },
  }),
  '/download/:downloadItemId',
);
