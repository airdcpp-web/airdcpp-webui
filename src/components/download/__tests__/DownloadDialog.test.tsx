import {
  getConnectedSocket,
  getMockServer,
  waitForExpect,
} from 'airdcpp-apisocket/tests/helpers.js';

import { jest } from '@jest/globals';
import ShareConstants from 'constants/ShareConstants';
import { ShareGetGroupedRootsResponse } from 'tests/mocks/api/share';
import FavoriteDirectoryConstants from 'constants/FavoriteDirectoryConstants';
import { FavoriteDirectoriesGroupedPathsResponse } from 'tests/mocks/api/favorite-directories';
import HistoryConstants, { HistoryStringEnum } from 'constants/HistoryConstants';
import {
  FilelistGetFilelistItemResponse,
  FilelistGetResponse,
  MOCK_FILELIST_ITEM_ID,
} from 'tests/mocks/api/filelist';
import FilesystemConstants from 'constants/FilesystemConstants';
import {
  FilesystemDiskInfoResponse,
  FilesystemListContentResponse,
} from 'tests/mocks/api/filesystem';
import { renderRoutes } from 'tests/test-containers';
import DownloadDialog from '../DownloadDialog';

import * as UI from 'types/ui';
import * as API from 'types/api';
import { MockHintedUserResponse } from 'tests/mocks/api/user';
import {
  act,
  waitForElementToBeRemoved,
  waitFor,
  fireEvent,
} from '@testing-library/react';

// import preview from 'jest-preview';
import { HistoryStringPathResponse } from 'tests/mocks/api/history';
import {
  createTestModalController,
  TestModalNavigateButton,
  waitForData,
} from 'tests/test-component-helpers';

// tslint:disable:no-empty
describe('DownloadDialog', () => {
  let server: ReturnType<typeof getMockServer>;
  const getSocket = async () => {
    const { socket } = await getConnectedSocket(server);

    server.addDataHandler(
      'GET',
      ShareConstants.GROUPED_ROOTS_GET_URL,
      ShareGetGroupedRootsResponse,
    );
    server.addDataHandler(
      'GET',
      FavoriteDirectoryConstants.GROUPED_DIRECTORIES_URL,
      FavoriteDirectoriesGroupedPathsResponse,
    );
    server.addDataHandler(
      'GET',
      `${HistoryConstants.STRINGS_URL}/${HistoryStringEnum.DOWNLOAD_DIR}`,
      HistoryStringPathResponse,
    );

    server.addDataHandler(
      'POST',
      FilesystemConstants.DISK_INFO_URL,
      FilesystemDiskInfoResponse,
    );
    server.addDataHandler(
      'POST',
      FilesystemConstants.LIST_URL,
      FilesystemListContentResponse,
    );

    return socket;
  };

  const renderDialog = async () => {
    const socket = await getSocket();
    const handleDownload = jest.fn<UI.DownloadHandler<API.FilelistItem>>();

    const DownloadDialogTest = () => {
      return (
        <>
          <TestModalNavigateButton
            modalRoute={`/home/download/${MOCK_FILELIST_ITEM_ID}`}
          />
          <DownloadDialog
            downloadHandler={handleDownload}
            itemDataGetter={() =>
              Promise.resolve(FilelistGetFilelistItemResponse as API.FilelistItem)
            }
            userGetter={() => MockHintedUserResponse as API.HintedUser}
            session={FilelistGetResponse}
          />
        </>
      );
    };

    const routes = [
      {
        path: '/home/*',
        Component: DownloadDialogTest,
      },
    ];

    const renderData = await renderRoutes(routes, {
      socket,
      routerProps: { initialEntries: ['/home'] },
    });

    const modalController = createTestModalController(renderData);
    return { socket, handleDownload, modalController, ...renderData };
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  test('should load and close', async () => {
    const { getByText, socket, modalController } = await renderDialog();

    await modalController.openDialog();

    // Check content
    await waitForExpect(() => expect(getByText('Download')).toBeTruthy());

    await modalController.closeDialogButton('Close');

    socket.disconnect();
  });

  test('should handle download', async () => {
    const { socket, modalController, handleDownload } = await renderDialog();

    await modalController.openDialog();

    // Download
    await modalController.closeDialogText(HistoryStringPathResponse[0]);

    // Check
    expect(handleDownload).toHaveBeenCalledTimes(1);
    expect(handleDownload.mock.calls).toMatchSnapshot();

    socket.disconnect();
  });

  test('should open and close download dialog', async () => {
    const { socket, modalController, getByText, queryByText } = await renderDialog();

    // Open download file browser
    await modalController.openDialog();
    expect(fireEvent.click(getByText('Browse'))).toBeTruthy();

    // Open browse dialog
    await waitFor(() => queryByText(/Loading items/i));
    await waitForElementToBeRemoved(() => queryByText(/Loading items/i));

    // Close it
    await modalController.closeDialogButton('Cancel');

    socket.disconnect();
  });

  test('should handle browse dialog download', async () => {
    const { socket, modalController, handleDownload, getByText, queryByText } =
      await renderDialog();

    // Open download file browser
    await modalController.openDialog();
    expect(fireEvent.click(getByText('Browse'))).toBeTruthy();

    // Open browse dialog
    await waitForData('Loading items', queryByText);

    // Download
    await modalController.closeDialogButton('Download');

    expect(handleDownload).toHaveBeenCalledTimes(1);
    expect(handleDownload.mock.calls).toMatchSnapshot();

    socket.disconnect();
  });
});
