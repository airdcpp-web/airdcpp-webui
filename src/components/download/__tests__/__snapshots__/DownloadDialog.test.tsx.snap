// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadDialog should handle browse dialog download 1`] = `
[
  [
    {
      "complete": true,
      "dupe": null,
      "id": 323,
      "name": "ubuntu-20.04.2.0-desktop-amd64.iso",
      "path": "/Downloads/ubuntu-20.04.2.0-desktop-amd64.iso",
      "size": 2877227008,
      "time": 1618062636,
      "tth": "2E2P2K5JP6KARHNFGUMIFENN4SGBENKLPGQNUAA",
      "type": {
        "content_type": "",
        "id": "file",
        "str": "iso",
      },
    },
    {
      "cid": "XPI67YM335U72FUS626M6WYFGR5EBZTNCEJU3JA",
      "flags": [
        "ccpm",
        "asch",
      ],
      "hub_names": "Test hub 1 (Test hub 2)",
      "hub_url": "adcs://oden.dyndns.info:41214/?kp=SHA256/RW37PVLRXU5HEXXIKBCFVIKRI6I5T7RXUTFDG2PXBHVUNX4CKW5A",
      "hub_urls": [
        "adcs://test1.com:53245?kp=SHA256/EHNK6GWENKR7XBEMZPBDRZMENEIIGZV6NTPUZ2MJOCRH7TAPVOBQ",
        "adcs://test2.dyndns.info:5555",
      ],
      "nicks": "test1 ([100]test1)",
    },
    {
      "priority": -1,
      "target_directory": "/mnt/disk1/Images/",
      "target_name": "ubuntu-20.04.2.0-desktop-amd64.iso",
    },
    {
      "id": "HPI67YG335U72FUS626M6WYFGR5EBZTNCEJU3JA",
      "location": null,
      "partial_list": true,
      "read": false,
      "share_profile": null,
      "state": {
        "id": "download_pending",
        "str": "Download pending",
        "time_finished": 0,
      },
      "total_files": 936066,
      "total_size": 133137630299481,
      "user": {
        "cid": "XPI67YM335U72FUS626M6WYFGR5EBZTNCEJU3JA",
        "flags": [
          "ccpm",
          "asch",
        ],
        "hub_names": "Test hub 1 (Test hub 2)",
        "hub_url": "adcs://oden.dyndns.info:41214/?kp=SHA256/RW37PVLRXU5HEXXIKBCFVIKRI6I5T7RXUTFDG2PXBHVUNX4CKW5A",
        "hub_urls": [
          "adcs://test1.com:53245?kp=SHA256/EHNK6GWENKR7XBEMZPBDRZMENEIIGZV6NTPUZ2MJOCRH7TAPVOBQ",
          "adcs://test2.dyndns.info:5555",
        ],
        "nicks": "test1 ([100]test1)",
      },
    },
  ],
]
`;

exports[`DownloadDialog should handle download 1`] = `
[
  [
    {
      "complete": true,
      "dupe": null,
      "id": 323,
      "name": "ubuntu-20.04.2.0-desktop-amd64.iso",
      "path": "/Downloads/ubuntu-20.04.2.0-desktop-amd64.iso",
      "size": 2877227008,
      "time": 1618062636,
      "tth": "2E2P2K5JP6KARHNFGUMIFENN4SGBENKLPGQNUAA",
      "type": {
        "content_type": "",
        "id": "file",
        "str": "iso",
      },
    },
    {
      "cid": "XPI67YM335U72FUS626M6WYFGR5EBZTNCEJU3JA",
      "flags": [
        "ccpm",
        "asch",
      ],
      "hub_names": "Test hub 1 (Test hub 2)",
      "hub_url": "adcs://oden.dyndns.info:41214/?kp=SHA256/RW37PVLRXU5HEXXIKBCFVIKRI6I5T7RXUTFDG2PXBHVUNX4CKW5A",
      "hub_urls": [
        "adcs://test1.com:53245?kp=SHA256/EHNK6GWENKR7XBEMZPBDRZMENEIIGZV6NTPUZ2MJOCRH7TAPVOBQ",
        "adcs://test2.dyndns.info:5555",
      ],
      "nicks": "test1 ([100]test1)",
    },
    {
      "priority": -1,
      "target_directory": "/home/airdcpp/Downloads/",
      "target_name": "ubuntu-20.04.2.0-desktop-amd64.iso",
    },
    {
      "id": "HPI67YG335U72FUS626M6WYFGR5EBZTNCEJU3JA",
      "location": null,
      "partial_list": true,
      "read": false,
      "share_profile": null,
      "state": {
        "id": "download_pending",
        "str": "Download pending",
        "time_finished": 0,
      },
      "total_files": 936066,
      "total_size": 133137630299481,
      "user": {
        "cid": "XPI67YM335U72FUS626M6WYFGR5EBZTNCEJU3JA",
        "flags": [
          "ccpm",
          "asch",
        ],
        "hub_names": "Test hub 1 (Test hub 2)",
        "hub_url": "adcs://oden.dyndns.info:41214/?kp=SHA256/RW37PVLRXU5HEXXIKBCFVIKRI6I5T7RXUTFDG2PXBHVUNX4CKW5A",
        "hub_urls": [
          "adcs://test1.com:53245?kp=SHA256/EHNK6GWENKR7XBEMZPBDRZMENEIIGZV6NTPUZ2MJOCRH7TAPVOBQ",
          "adcs://test2.dyndns.info:5555",
        ],
        "nicks": "test1 ([100]test1)",
      },
    },
  ],
]
`;
